contents/dart/Dart.md

```diff
@@ -0,0 +1,228 @@
+
+However, if you only want to request the content of `f2` after `f1` is ready, you may nest callback functions like below:
+
+```dart
+main() {
+  var f1 = Future.delayed(Duration(seconds: 1), () => 'String from the future');
+  var f2 = Future.delayed(Duration(seconds: 2), () => 'Must come second');
+  f1.then((res1) {
+    f2.then((res2) {
+      print(res1 + " " + res2);
+    });
+  });
+}
+```
+
+<panel header="DartPad example"><iframe style="width: 100%; height: 400px;" src="https://dartpad.dev/embed-inline.html?id=ff4a94d26a59c373b7cd94cd8df7094a&split=50"></iframe></panel>
+
+This time, after 2 seconds, two strings should be printed out at the same time. However, using `async-await`, there's a more elegant way to do the same thing:
+
+```dart
+main() {
+  var f1 = Future.delayed(Duration(seconds: 1), () => 'String from the future');
+  var f2 = Future.delayed(Duration(seconds: 2), () => 'Must come second');
+  f1.then((res1) async {
+    print(res1 + " " + await f2);
+  });
+}
+```
+<panel header="DartPad example"><iframe style="width: 100%; height: 400px;" src="https://dartpad.dev/embed-inline.html?id=077d288e4a705324c4aa268810904305&split=50"></iframe></panel>
+</panel>
+
+### Benefit: Extension methods
+Extension methods aim to solve one problem: when using an external library, you may want to extend or even change some of the methods for your own needs. For example, instead of using `int.parse(42)`, you want to extend the `String` class so that it has a method `parseInt()` to parse a string to int. You can easily do so by extending `String` class in Dart:
+
+```dart
+extension NumberParsing on String {
+  int parseInt() {
+    return int.parse(this);
+  }
+}
+
+main(List<String> args) {
+  print('42'.parseInt());
+}
+```
+
+<panel header="DartPad example"><iframe style="width: 100%; height: 400px;" src="https://dartpad.dev/embed-inline.html?id=c162704838776bb7da6568c3c1589f69&split=50"></iframe></panel>
+
+Extending classes, especially core classes like `String`, could be quite tricky if not impossible without extension methods. However, in Dart, you can easily customize the behavior of built-in classes like so. Extension methods in Dart have even more interesting stuff like [generic support](https://dart.dev/guides/language/extension-methods#implementing-generic-extensions).
+
+### Benefit: Named parameters
+This is what makes Dart code look like JavaScript code (or JSON-like, as mentioned before). In languages like Java, the order of a function's parameters is specified by the function's definition, and if you only look at the function call, it might be difficult for you to figure out what those parameters mean. One classic example is the `put(K key, V value)` method of  `Map<K, V>` interface in Java - if you only see `m.put(1, 2)`, can you tell which one is key without prior knowledge?
```

Comment:
Try to lower the necessary background knowledge to read this article in general; The reader might not know Java